src/xl_wrapper.py

```diff
@@ -186,136 +191,176 @@ def from_pretrained(cls, model_name_or_path=None, seq_len=512, weights_path=None
-    #             # Flatten the tokens
-    #             loss_fct = CrossEntropyLoss(ignore_index=self.pad_token_id)
-    #             loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
-    #         res.append((lm_logits, loss))
-    #     logits = torch.cat([x[0] for x in res], dim=0)[:, : original_context_length, :]
-    #     if loss is not None:
-    #         loss = [x[1] for x in res]
-    #     return ModelOutput(logits, loss)
+            while context_length > seq_len:
+                seq_len += 16
+            if context_length < seq_len:
+                context_tokens.extend([self.pad_token_id] * (seq_len - context_length))
+                if labels is not None:
+                    lbl = lbl.tolist()
+                    lbl.extend([self.pad_token_id] * (seq_len - context_length))
+                    lbl = torch.cuda.LongTensor(lbl)
+            if context_length > 2048:
+                context_tokens = context_tokens[-2048:]
+                if labels is not None:
+                    lbl = lbl.tolist()[-2048:]
+                    lbl = torch.cuda.LongTensor(lbl)
+            context_tokens_tensor = torch.cuda.LongTensor(context_tokens)
+            context_length_tensor = torch.cuda.LongTensor([context_length])
+
+            torch.distributed.broadcast(context_length_tensor, mpu.get_model_parallel_src_rank(),
+                                        group=mpu.get_model_parallel_group())
+            torch.distributed.broadcast(context_tokens_tensor, mpu.get_model_parallel_src_rank(),
+                                        group=mpu.get_model_parallel_group())
+
+            # context_length = context_length_tensor[0].item()
+
+            tokens = context_tokens_tensor
+            tokens = tokens.view(1, -1).contiguous()
+            tokens = tokens.to(torch.cuda.current_device())
+            attention_mask, loss_mask, position_ids = get_masks_and_position_ids(tokens, self.pad_token_id, False,
+                                                                                 False)
+            lm_logits = self.model(tokens, position_ids, attention_mask)
+            loss = None
+            if labels is not None:
+                # Shift so that tokens < n predict n
+                shift_logits = lm_logits[..., :-1, :].contiguous()
+                shift_labels = lbl[..., 1:].contiguous()
+                # Flatten the tokens
+                loss_fct = CrossEntropyLoss(ignore_index=self.pad_token_id)
+                loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
+            res.append((lm_logits, loss))
+        logits = torch.cat([x[0] for x in res], dim=0)[:, : original_context_length, :]
+        if loss is not None:
+            loss = [x[1] for x in res]
+        return ModelOutput(logits, loss)
+    
+    def resize_token_embeddings(self, new_num_tokens = None):    # -> nn.Embedding:
```

Comment:
after setting the input & output embeddings, this does the GPU allocation and Fp16 conversion again.